src/main/kotlin/org/rust/ide/folding/RustFoldingBuilder.kt

```diff
@@ -6,79 +6,139 @@ import com.intellij.lang.folding.FoldingDescriptor
             }
-        }
-        root.descendentsOfType<RustImplMethodMemberElement>().forEach {
-            val methodBody = it.block
-            if (methodBody != null) {
-                descriptors += FoldingDescriptor(it.node, methodBody.textRange)
+
+            override fun visitTraitItem(o: RustTraitItemElement) {
+                super.visitTraitItem(o)
+
+                val traitBody = o.traitBody
+                descriptors += FoldingDescriptor(o.node, traitBody.textRange)
             }
-        }
-        root.descendentsOfType<RustTraitMethodMemberElement>().forEach {
-            val methodBody = it.block
-            if (methodBody != null) {
-                descriptors += FoldingDescriptor(it.node, methodBody.textRange)
+
+            override fun visitEnumVariant(o: RustEnumVariantElement) {
+                super.visitEnumVariant(o)
+
+                val structDeclArgs = o.enumStructArgs
+                if (structDeclArgs != null) {
+                    descriptors += FoldingDescriptor(o.node, structDeclArgs.textRange)
+                }
             }
-        }
-        root.descendentsOfType<RustModItemElement>().forEach {
-            val rbrace = it.rbrace;
-            if (rbrace != null) {
-                descriptors += FoldingDescriptor(it.node, TextRange(it.lbrace.textOffset, rbrace.textOffset + 1))
+
+            override fun visitFn(o: RustFnElement) {
+                super.visitFn(o)
+
+                val fnBody = o.block
+                if (fnBody != null) {
+                    descriptors += FoldingDescriptor(o.node, fnBody.textRange)
+                }
             }
-        }
+
+            override fun visitModItem(o: RustModItemElement) {
+                super.visitModItem(o)
+
+                val rbrace = o.rbrace;
+                if (rbrace != null) {
+                    descriptors += FoldingDescriptor(o.node, TextRange(o.lbrace.textOffset, rbrace.textOffset + 1))
+                }
+            }
+
+            override fun visitMatchExpr(o: RustMatchExprElement) {
+                super.visitMatchExpr(o)
+
+                val body = o.matchBody
+                if (body != null) {
+                    descriptors += FoldingDescriptor(o.node, body.textRange)
+                }
+            }
+
+            override fun visitMacroArg(o: RustMacroArgElement) {
+                super.visitMacroArg(o)
+
+                val lbrace = o.lbrace
+                val rbrace = o.rbrace
+
+                if (lbrace != null && rbrace != null) {
+                    descriptors += FoldingDescriptor(o.node, TextRange(lbrace.textOffset, rbrace.textOffset + 1))
+                }
+            }
+
+            override fun visitUseGlobList(o: RustUseGlobListElement) {
+                super.visitUseGlobList(o)
+
+                descriptors += FoldingDescriptor(o.node, o.textRange)
+            }
+
+            override fun visitComment(comment: PsiComment?) {
```

Comment:
You can remove `?` from the method signature. It is added by IDE just because the parent method is not annotated, not because it can really be null: https://youtrack.jetbrains.com/issue/KT-12213